cleanupBuckets.js

```diff
@@ -0,0 +1,160 @@
+const http = require('http');
+const async = require('async');
+const AWS = require('aws-sdk');
+const { Logger } = require('werelogs');
+
+const log = new Logger('s3utils::emptyBucket');
+// configurable params
+const BUCKETS = process.argv[2] ? process.argv[2].split(',') : null;
+const ACCESS_KEY = process.env.ACCESS_KEY;
+const SECRET_KEY = process.env.SECRET_KEY;
+const ENDPOINT = process.env.endpoint;
+if (!BUCKETS || BUCKETS.length === 0) {
+    log.fatal('No buckets given as input! Please provide ' +
+        'a comma-separated list of buckets');
+    process.exit(1);
+}
+if (!ENDPOINT) {
+    log.fatal('ENDPOINT not defined!');
+    process.exit(1);
+}
+if (!ACCESS_KEY) {
+    log.fatal('ACCESS_KEY not defined');
+    process.exit(1);
+}
+if (!SECRET_KEY) {
+    log.fatal('SECRET_KEY not defined');
+    process.exit(1);
+}
+const LISTING_LIMIT = 300;
+
+AWS.config.update({
+    accessKeyId: ACCESS_KEY,
+    secretAccessKey: SECRET_KEY,
+    endpoint: ENDPOINT,
+    region: 'us-east-1',
+    sslEnabled: false,
+    s3ForcePathStyle: true,
+    apiVersions: { s3: '2006-03-01' },
+    signatureVersion: 'v4',
+    signatureCache: false,
+});
+
+const s3 = new AWS.S3({
+    httpOptions: {
+        maxRetries: 0,
+        timeout: 0,
+        agent: new http.Agent({ keepAlive: true }),
+    },
+});
+
+// list object versions
+function _listObjectVersions(bucket, VersionIdMarker, KeyMarker, cb) {
+    return s3.listObjectVersions({
+        Bucket: bucket,
+        MaxKeys: LISTING_LIMIT,
+        VersionIdMarker,
+        KeyMarker,
+    }, cb);
+}
+
+// return object with key and version_id
+function _getKeys(keys) {
+    return keys.map(v => ({
+        Key: v.Key,
+        VersionId: v.VersionId,
+    }));
+}
+
+// delete all versions of an object
+function _deleteVersions(bucket, objectsToDelete, cb) {
+    // multi object delete can delete max 1000 objects
+    function _batchDelete(Objects, done) {
+        const params = {
+            Bucket: bucket,
+            Delete: { Objects },
+        };
+        s3.deleteObjects(params, err => {
+            if (err) {
+                log.error('batch delete err', err);
+                return done(err);
+            }
+            Objects.forEach(v => log.info(`deleted key: ${v.Key}`));
+            return done();
+        });
+    }
+    const fns = [];
+    while (Object.keys(objectsToDelete).length > 0) {
```

Comment:
I'm not sure if this is needed because I don't see `objectsToDelete` ever being greater than 1000

Reply:
Correct. It's no longer the case that listing returns results more than 1000.